email_backend/data_management/generate_csv.js

```diff
@@ -0,0 +1,92 @@
+import { connect_db } from "../src/db_utils/connect_db";
+import { get_templates } from "../src/template_utils";
+import { make_mongoose_model_from_original_template } from "../src/db_utils/log_email_and_meta_to_db";
+import _ from "lodash";
+import fs from "fs";
+import mongoose from "mongoose";
+import { Parser } from "json2csv";
+
+const json2csv_parser = new Parser();
+
+connect_db()
+  .then(() => {
+    const templates = get_templates();
+    let templates_left = _.keys(templates).length;
+    _.forEach(templates, (template_value, template_name) => {
+      const collection = make_mongoose_model_from_original_template({
+        original_template: template_value,
+        template_name,
+      });
+
+      //if it happens that later on we add another key with a json value,
+      //change reduced json to handle "key.value". For the time being, the CSV
+      //will be cleaner if we flatten the "key.value" to "value"
+
+      collection.find({}, function (err, email_logs) {
```

Comment:
Aha, no need for the `templates_left` counter, this is a problem to solve with promises! Cleanup like closing connections should go in a `.finally()` at the end of the promises chain (after the `.catch`, same as putting cleanup code after a try-catch block).

Reply:
While I'm giving promise-advice, `Model.find` only takes a callback function for legacy reasons (callback hell, the pre-promises state of all async JS code), but if you call it as `Model.find(...).exec()` then the return value is a promise for the (eventually) resolved query. Much nicer to handle than callbacks!

You can keep a promise chain going if you return another promise from a `.then()`. The next `.then()` will receive the result of the previous promise as an argument, once it resolves. In this case, you'd probably want this first `.then` to return a promise for all the async `.find` calls resolving, and then handle the data processing in an additional `.then` call in the chain.

The `Promise` object itself has a couple methods for handling the resolution of multiple promises:`all`, `allSettled`, `any`, and `race`. The differences between some of these are subtle, read the MDN Promises docs for more details. The point here is that each step in a promise chain can ultimately only return one promise (or a non-promise value in the last step), so you'll need to wrap all of the promises from the data fetching step in a `Promises.all` to collapse them to a single promise.

I.e. (untested code, no guarantee I didn't typo something here or miss something)
```
connect_db()
  .then(() => _.chain(get_templates())
    .map((template_value, template_name) => make_mongoose_model_from_original_template( /* stuff */))
    .map((model) => model.find({}).exec())
    .thru(Promises.all)
    .value()
  )
  .then((all_email_logs) => {
    // ... process data
  })
  .catch( (error) => console.log(error) )
  .finally( () => mongoose.connection.close() )
```

Note that if any of the promises from the `model.find({}).exec()` calls reject/throw errors than the data processing step never happens. This may or may not be the desired behaviour. I think it makes sense though, keeps the whole script all-or-nothing which is generally a better way to be (even if it might be annoying if one broken template stops you from getting the data from the others).